src/nvim/api/vim.c

```diff
@@ -2276,6 +2276,12 @@ Dictionary nvim_eval_statusline(String str, Dict(eval_statusline) *opts, Error *
   bool use_tabline = false;
   bool highlights = false;
 
+  const char *const errmsg = check_stl_option((char_u *)str.data);
+  if (errmsg) {
+    api_set_error(err, kErrorTypeValidation, "%s", errmsg);
+    return result;
+  }
+
```

Comment:
You need to make an exception for when the string starts with `%!`, see:

https://github.com/neovim/neovim/blob/df4bdba58fed4d88681fe71ce644c42d5578613d/src/nvim/option.c#L2944-L2945

```suggestion
  if (str.size < 2 || memcmp(str.data, "%!", 2)) {
    const char *const errmsg = check_stl_option((char_u *)str.data);
    if (errmsg) {
      api_set_error(err, kErrorTypeValidation, "%s", errmsg);
      return result;
    }
  }
```